application_sdk/common/observability.py

```diff
@@ -0,0 +1,202 @@
+                "level": message.record["level"].name,
+                "logger_name": message.record["extra"].get("logger_name", ""),
+                "message": message.record["message"],
+                "file": str(message.record["file"].path),
+                "line": message.record["line"],
+                "function": message.record["function"],
+            }
+            for key, value in message.record["extra"].items():
+                if key not in log_record:
+                    log_record[key] = value
+
+            with self._buffer_lock:
+                self._log_buffer.append(log_record)
+                now = time()
+                if (
+                    len(self._log_buffer) >= self._batch_size
+                    or (now - self._last_flush_time) >= self._flush_interval
+                ):
+                    self._last_flush_time = now
+                    buffer_copy = self._log_buffer[:]
+                    self._log_buffer.clear()
+                else:
+                    buffer_copy = None
+
+            if buffer_copy is not None:
+                await self._flush_records(buffer_copy)
+        except Exception as e:
+            logging.error(f"Error buffering log: {e}")
+
+    async def _flush_buffer(self, force=False):
+        with self._buffer_lock:
+            if self._log_buffer:
+                buffer_copy = self._log_buffer[:]
+                self._log_buffer.clear()
+            else:
+                buffer_copy = None
+        if buffer_copy:
+            await self._flush_records(buffer_copy)
+
+    async def _flush_records(self, records):
+        try:
+            df = pd.DataFrame(records)
+            if os.path.exists(self.parquet_path):
+                existing_table = pq.read_table(self.parquet_path)
+                existing_df = existing_table.to_pandas()
+                df = pd.concat([existing_df, df], ignore_index=True)
+            table = pa.Table.from_pandas(df)
+            pq.write_table(table, self.parquet_path, compression="snappy")
+            with open(self.parquet_path, "rb") as f:
+                file_content = f.read()
+                filename = "logs/log.parquet"
+                metadata = {
+                    "key": filename,
+                    "blobName": filename,
+                    "fileName": filename,
+                }
+                with DaprClient() as client:
+                    client.invoke_binding(
+                        binding_name=OBJECT_STORE_NAME,
+                        operation="create",
+                        data=file_content,
+                        binding_metadata=metadata,
+                    )
+            # Clean up old logs after flushing
+            if LOG_CLEANUP_ENABLED:
+                await self._cleanup_old_logs()
+        except Exception as e:
+            logging.error(f"Error flushing log batch to parquet: {e}")
+
+    async def _cleanup_old_logs(self):
```

Comment:
a little big for cleanup function, can we breakdown if needed post removing dapr operations (via re-use)?